File: sdk/python-v2/src/openwop_client/webhook_helpers.py
"""Webhook delivery-verification helpers per ``spec/v2/core/webhooks.md``
§Verification.

Receivers MUST verify both the HMAC AND the timestamp freshness before
accepting a delivery — verifying HMAC alone leaves the receiver open to
replay attacks.

The signing recipe (webhooks.md §Headers)::

    hmac = HMAC-SHA256(secret, f"{timestamp}.{rawBody}")
    header OpenWOP-Signature:           sha256=<hmac-hex>
    header OpenWOP-Timestamp:           <unix-seconds>
    header OpenWOP-Signature-Algorithm: v1

A host advertising both majors sends the ``X-openwop-*`` family beside it with
identical values through the overlap; :func:`read_webhook_headers` accepts
either family. The SDK-only ``openwop-Webhook-*`` names and the ``v1=<hex>``
value form were removed in v2 (headers.md §Removed).

Verification:

    1. Parse the ``sha256=<hex>`` value from the signature header.
    2. Reject an unrecognized ``OpenWOP-Signature-Algorithm`` value (MUST).
    3. Reject a timestamp more than ±window from the clock (default 5 minutes).
    4. Recompute ``HMAC-SHA256(f"{timestamp}.{rawBody}", secret)`` and compare
       in constant time (``hmac.compare_digest``).

The HMAC is computed over the raw body bytes; receivers MUST pass the exact
bytes the host POSTed, NOT a re-serialized parsed representation.
"""

from __future__ import annotations

import hmac
import re
import time
from dataclasses import dataclass
from typing import Literal, Mapping, Union

DEFAULT_WEBHOOK_FRESHNESS_WINDOW_SECONDS = 300

#: The one signature scheme (``OpenWOP-Signature-Algorithm``): HMAC-SHA256 with
#: the subscription secret.
WEBHOOK_SIGNATURE_ALGORITHMS: frozenset[str] = frozenset({"v1"})

VerifyReason = Literal[
    "signature_mismatch",
    "timestamp_expired",
    "timestamp_too_far_in_future",
    "malformed_signature_header",
    "malformed_timestamp_header",
    "unsupported_signature_algorithm",
]


@dataclass(frozen=True)
class VerifyValid:
    valid: Literal[True] = True


@dataclass(frozen=True)
class VerifyInvalid:
    reason: VerifyReason
    valid: Literal[False] = False


VerifyWebhookOutcome = Union[VerifyValid, VerifyInvalid]


_HEX_RE = re.compile(r"^[0-9a-fA-F]+$")
_SIGNATURE_VALUE_PREFIX = "sha256="


def verify_webhook_signature(
    secret: str,
    signature_header: str,
    timestamp_header: str,
    raw_body: bytes | str,
    *,
    freshness_window_seconds: int = DEFAULT_WEBHOOK_FRESHNESS_WINDOW_SECONDS,
    now_seconds: int | None = None,
    algorithm_header: str | None = None,
) -> VerifyWebhookOutcome:
    """Verify a webhook delivery.

    Returns :class:`VerifyValid` on success, :class:`VerifyInvalid` otherwise
    with a ``reason`` field for diagnostics. ``algorithm_header`` is the
    ``OpenWOP-Signature-Algorithm`` value when the caller read one; an
    unrecognized value is rejected.
    """

    provided_hex = parse_signature_value(signature_header)
    if provided_hex is None:
        return VerifyInvalid(reason="malformed_signature_header")

    if algorithm_header is not None and algorithm_header not in WEBHOOK_SIGNATURE_ALGORITHMS:
        return VerifyInvalid(reason="unsupported_signature_algorithm")

    try:
        timestamp = int(timestamp_header)
    except (ValueError, TypeError):
        return VerifyInvalid(reason="malformed_timestamp_header")
    if timestamp <= 0:
        return VerifyInvalid(reason="malformed_timestamp_header")

    if freshness_window_seconds > 0:
        now = now_seconds if now_seconds is not None else int(time.time())
        delta = now - timestamp
        if delta > freshness_window_seconds:
            return VerifyInvalid(reason="timestamp_expired")
        if delta < -freshness_window_seconds:
            return VerifyInvalid(reason="timestamp_too_far_in_future")

    body_bytes = raw_body if isinstance(raw_body, bytes) else raw_body.encode("utf-8")
    signed = f"{timestamp}.".encode("utf-8") + body_bytes
    expected_hex = hmac.new(secret.encode("utf-8"), signed, "sha256").hexdigest()

    if not hmac.compare_digest(provided_hex.lower(), expected_hex.lower()):
        return VerifyInvalid(reason="signature_mismatch")
    return VerifyValid()


def sign_webhook_delivery(
    secret: str,
    timestamp: int,
    raw_body: bytes | str,
) -> tuple[str, str]:
    """Compute the canonical webhook signature for a payload.

    Returns ``(signature_header, timestamp_header)`` — ``sha256=<hex>`` and the
    timestamp — useful when implementing a host or generating test fixtures.
    See :func:`webhook_delivery_headers` for the full header set.
    """

    body_bytes = raw_body if isinstance(raw_body, bytes) else raw_body.encode("utf-8")
    signed = f"{timestamp}.".encode("utf-8") + body_bytes
    hex_digest = hmac.new(secret.encode("utf-8"), signed, "sha256").hexdigest()
    return f"sha256={hex_digest}", str(timestamp)


def webhook_delivery_headers(
    secret: str,
    timestamp: int,
    raw_body: bytes | str,
) -> dict[str, str]:
    """The signature headers a v2 host sends on a delivery: the ``OpenWOP-*``
    family plus the ``X-openwop-*`` twins a dual-major host emits through the
    overlap (webhooks.md §Dual emission). ``OpenWOP-Webhook-Id`` and
    ``OpenWOP-Event-Type`` are subscription-specific and added by the host."""
    sig, ts = sign_webhook_delivery(secret, timestamp, raw_body)
    return {
        "OpenWOP-Signature": sig,
        "OpenWOP-Timestamp": ts,
        "OpenWOP-Signature-Algorithm": "v1",
        "X-openwop-Signature": sig,
        "X-openwop-Timestamp": ts,
        "X-openwop-Signature-Algorithm": "v1",
    }


def parse_signature_value(value: str) -> str | None:
    """``sha256=<hex>`` -> ``<hex>``; anything else (including the removed
    ``v1=<hex>`` form) -> None."""
    if not value.startswith(_SIGNATURE_VALUE_PREFIX):
        return None
    hex_part = value[len(_SIGNATURE_VALUE_PREFIX):]
    return hex_part if _HEX_RE.fullmatch(hex_part) else None


#: Receiver preference order: the v2 ``OpenWOP-*`` family, then the
#: ``X-openwop-*`` family accepted through the overlap. Each row is
#: ``(signature_header, timestamp_header, algorithm_header, family)``.
#: Lookups are case-insensitive.
WEBHOOK_HEADER_FAMILIES: tuple[tuple[str, str, str, str], ...] = (
    ("OpenWOP-Signature", "OpenWOP-Timestamp", "OpenWOP-Signature-Algorithm", "openwop"),
    ("X-openwop-Signature", "X-openwop-Timestamp", "X-openwop-Signature-Algorithm", "x-openwop"),
)


@dataclass(frozen=True)
class WebhookHeaderRead:
    signature: str
    timestamp: str
    family: Literal["openwop", "x-openwop"]
    #: The ``*-Signature-Algorithm`` value when the delivery carried one.
    algorithm: str | None = None


def read_webhook_headers(headers: "Mapping[str, str]") -> WebhookHeaderRead | None:
    """Return the first complete header family's values, or None. ``headers`` is
    any mapping (case-insensitive or not)."""
    lowered = {k.lower(): v for k, v in headers.items()}
    for sig_name, ts_name, alg_name, family in WEBHOOK_HEADER_FAMILIES:
        sig = lowered.get(sig_name.lower())
        ts = lowered.get(ts_name.lower())
        if sig is not None and ts is not None:
            fam: Literal["openwop", "x-openwop"] = "openwop" if family == "openwop" else "x-openwop"
            return WebhookHeaderRead(
                signature=sig,
                timestamp=ts,
                family=fam,
                algorithm=lowered.get(alg_name.lower()),
            )
    return None

File: go/v2/webhook_helpers.go
// Webhook delivery-verification helpers per spec/v2/core/webhooks.md
// §Verification. Receivers MUST verify both the HMAC AND the timestamp
// freshness before accepting a delivery — verifying HMAC alone leaves the
// receiver open to replay attacks.
//
// The signing recipe (webhooks.md §Headers):
//
//	hmac = HMAC-SHA256(secret, fmt.Sprintf("%d.%s", timestamp, rawBody))
//	header OpenWOP-Signature:           sha256=<hmac-hex>
//	header OpenWOP-Timestamp:           <unix-seconds>
//	header OpenWOP-Signature-Algorithm: v1
//
// A host advertising both majors sends the X-openwop-* family beside it with
// identical values through the overlap; ReadWebhookHeaders accepts either
// family. The SDK-only openwop-Webhook-* names and the v1=<hex> value form
// were removed in v2 (headers.md §Removed).
//
// Verification:
//
//  1. Parse the sha256=<hex> value from the signature header.
//  2. Reject an unrecognized OpenWOP-Signature-Algorithm value (MUST).
//  3. Reject a timestamp more than ±window from the clock (default 5 min).
//  4. Recompute HMAC-SHA256 over "<timestamp>.<rawBody>" and compare with
//     constant-time hmac.Equal.

package openwopclient

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"strconv"
	"strings"
	"time"
)

// DefaultWebhookFreshnessWindowSeconds is the default replay-attack
// freshness window per webhooks.md §Verification (±5 minutes).
const DefaultWebhookFreshnessWindowSeconds = 300

// WebhookSignatureAlgorithms lists the recognized OpenWOP-Signature-Algorithm
// values: "v1" is HMAC-SHA256 with the subscription secret.
var WebhookSignatureAlgorithms = []string{"v1"}

// VerifyWebhookOutcome is the result of VerifyWebhookSignature.
// Valid is true on success; Reason carries the failure mode otherwise.
type VerifyWebhookOutcome struct {
	Valid  bool
	Reason VerifyWebhookReason
}

// VerifyWebhookReason names the specific failure mode when Valid is false.
type VerifyWebhookReason string

const (
	VerifyReasonSignatureMismatch       VerifyWebhookReason = "signature_mismatch"
	VerifyReasonTimestampExpired        VerifyWebhookReason = "timestamp_expired"
	VerifyReasonTimestampTooFarInFuture VerifyWebhookReason = "timestamp_too_far_in_future"
	VerifyReasonMalformedSigHeader      VerifyWebhookReason = "malformed_signature_header"
	VerifyReasonMalformedTSHeader       VerifyWebhookReason = "malformed_timestamp_header"
	VerifyReasonUnsupportedAlgorithm    VerifyWebhookReason = "unsupported_signature_algorithm"
)

// VerifyWebhookOptions overrides the freshness window or the wall clock for
// testing, and carries the algorithm header when the caller read one.
// Zero-valued fields use the defaults.
type VerifyWebhookOptions struct {
	// FreshnessWindowSeconds caps the age of accepted deliveries.
	// Zero means "use default" (300 seconds); -1 disables the
	// timestamp check (NOT recommended).
	FreshnessWindowSeconds int
	// NowSeconds overrides time.Now().Unix() for testing.
	NowSeconds int64
	// AlgorithmHeader is the OpenWOP-Signature-Algorithm value the delivery
	// carried (ReadWebhookHeaders returns it). Empty = not checked; an
	// unrecognized value is rejected.
	AlgorithmHeader string
}

// VerifyWebhookSignature verifies a webhook delivery. Returns Valid=true on
// success; otherwise Valid=false with Reason set so callers can log + alert.
//
// Callers MUST pass the raw body bytes — re-serialized parsed JSON will fail
// verification because the host signs exact bytes.
func VerifyWebhookSignature(
	secret string,
	signatureHeader string,
	timestampHeader string,
	rawBody []byte,
	opts VerifyWebhookOptions,
) VerifyWebhookOutcome {
	providedHex, ok := ParseSignatureValue(signatureHeader)
	if !ok {
		return VerifyWebhookOutcome{Valid: false, Reason: VerifyReasonMalformedSigHeader}
	}
	providedBytes, err := hex.DecodeString(providedHex)
	if err != nil || len(providedBytes) == 0 {
		return VerifyWebhookOutcome{Valid: false, Reason: VerifyReasonMalformedSigHeader}
	}

	if opts.AlgorithmHeader != "" && !isWebhookSignatureAlgorithm(opts.AlgorithmHeader) {
		return VerifyWebhookOutcome{Valid: false, Reason: VerifyReasonUnsupportedAlgorithm}
	}

	timestamp, err := strconv.ParseInt(timestampHeader, 10, 64)
	if err != nil || timestamp <= 0 {
		return VerifyWebhookOutcome{Valid: false, Reason: VerifyReasonMalformedTSHeader}
	}

	window := opts.FreshnessWindowSeconds
	if window == 0 {
		window = DefaultWebhookFreshnessWindowSeconds
	}
	if window > 0 {
		now := opts.NowSeconds
		if now == 0 {
			now = time.Now().Unix()
		}
		delta := now - timestamp
		if delta > int64(window) {
			return VerifyWebhookOutcome{Valid: false, Reason: VerifyReasonTimestampExpired}
		}
		if delta < -int64(window) {
			return VerifyWebhookOutcome{Valid: false, Reason: VerifyReasonTimestampTooFarInFuture}
		}
	}

	mac := hmac.New(sha256.New, []byte(secret))
	mac.Write([]byte(fmt.Sprintf("%d.", timestamp)))
	mac.Write(rawBody)
	expected := mac.Sum(nil)

	if !hmac.Equal(providedBytes, expected) {
		return VerifyWebhookOutcome{Valid: false, Reason: VerifyReasonSignatureMismatch}
	}
	return VerifyWebhookOutcome{Valid: true}
}

func isWebhookSignatureAlgorithm(value string) bool {
	for _, alg := range WebhookSignatureAlgorithms {
		if value == alg {
			return true
		}
	}
	return false
}

// SignWebhookDelivery computes the canonical webhook signature for a
// payload — useful when implementing a host or generating test fixtures.
// Returns (signatureHeader, timestampHeader): "sha256=<hex>" and the
// timestamp. See WebhookDeliveryHeaders for the full header set.
func SignWebhookDelivery(secret string, timestamp int64, rawBody []byte) (string, string) {
	mac := hmac.New(sha256.New, []byte(secret))
	mac.Write([]byte(fmt.Sprintf("%d.", timestamp)))
	mac.Write(rawBody)
	return "sha256=" + hex.EncodeToString(mac.Sum(nil)), strconv.FormatInt(timestamp, 10)
}

// WebhookDeliveryHeaders returns the signature headers a v2 host sends on a
// delivery: the OpenWOP-* family plus the X-openwop-* twins a dual-major host
// emits through the overlap (webhooks.md §Dual emission). OpenWOP-Webhook-Id
// and OpenWOP-Event-Type are subscription-specific and added by the host.
func WebhookDeliveryHeaders(secret string, timestamp int64, rawBody []byte) map[string]string {
	sig, ts := SignWebhookDelivery(secret, timestamp, rawBody)
	return map[string]string{
		"OpenWOP-Signature":             sig,
		"OpenWOP-Timestamp":             ts,
		"OpenWOP-Signature-Algorithm":   "v1",
		"X-openwop-Signature":           sig,
		"X-openwop-Timestamp":           ts,
		"X-openwop-Signature-Algorithm": "v1",
	}
}

// ParseSignatureValue accepts the one v2 form "sha256=<hex>" (webhooks.md
// §Headers) and returns the hex digest. Anything else — including the
// removed "v1=<hex>" form — is malformed.
func ParseSignatureValue(value string) (string, bool) {
	const prefix = "sha256="
	if !strings.HasPrefix(value, prefix) {
		return "", false
	}
	h := value[len(prefix):]
	if h == "" {
		return "", false
	}
	if _, err := hex.DecodeString(h); err != nil {
		return "", false
	}
	return h, true
}

// WebhookHeaderFamily names the header triple a delivery carries.
type WebhookHeaderFamily struct {
	Signature string
	Timestamp string
	Algorithm string
	Family    string // "openwop" or "x-openwop"
}

// WebhookHeaderFamilies is the receiver's preference order: the v2 OpenWOP-*
// family, then the X-openwop-* family accepted through the overlap.
var WebhookHeaderFamilies = []WebhookHeaderFamily{
	{Signature: "OpenWOP-Signature", Timestamp: "OpenWOP-Timestamp", Algorithm: "OpenWOP-Signature-Algorithm", Family: "openwop"},
	{Signature: "X-openwop-Signature", Timestamp: "X-openwop-Timestamp", Algorithm: "X-openwop-Signature-Algorithm", Family: "x-openwop"},
}

// ReadWebhookHeaders picks the signature, timestamp and (optional) algorithm
// values from a delivery's headers — first complete family wins; lookups are
// case-insensitive via the getter, e.g. http.Header.Get. ok is false when no
// family is complete.
func ReadWebhookHeaders(get func(name string) string) (signature, timestamp, algorithm, family string, ok bool) {
	for _, f := range WebhookHeaderFamilies {
		sig := get(f.Signature)
		ts := get(f.Timestamp)
		if sig != "" && ts != "" {
			return sig, ts, get(f.Algorithm), f.Family, true
		}
	}
	return "", "", "", "", false
}
